### ISSUE TITLE
Connecting does not work after v0.6.3

### ISSUE THREAD

**Issue by sindrenm** [socketio/socket.io-client-java#352]
Hi,

I have been debugging this issue for a while now, but I can't seem to figure out what's wrong. We're migrating from [Java-WebSocket] (and a bare-bones WebSocket server) to socket.io, and it's working fine for our web clients. However, I'm having some issues with the Android client ((using this client library) – it never receives neither connect event (`EVEN_CONNECT`, `EVENT_CONNECTING`, `EVENT_CONNECT_ERROR`, `EVENT_CONNET_TIMEOUT`) when I connect to the socket.io server. I've tested earlier versions, and found that it stopped working in v0.7.0 (v6.3.0 worked fine). I'm getting the following log output on the server when I connect, so it seems as if it connects properly.

```
socket.io:server incoming connection with id bmo0ZIdaMqHIhGE8AAAA +14s
socket.io:client connecting to namespace / +1ms
socket.io:namespace adding socket to nsp / +0ms
socket.io:socket socket connected - writing packet +2ms
socket.io:socket joining room bmo0ZIdaMqHIhGE8AAAA +2ms
socket.io:client writing packet {"type":0,"nsp":"/"} +0ms
socket.io-parser encoding packet {"type":0,"nsp":"/"} +0ms
socket.io-parser encoded {"type":0,"nsp":"/"} as 0 +0ms
socket.io:socket joined room bmo0ZIdaMqHIhGE8AAAA +1ms
```

[java-websocket]: https://github.com/TooTallNate/Java-WebSocket

Here's the code I'm using on the Android client. I'm using Kotlin, but that shouldn't be an issue since that worked with v0.6.3.

```kotlin
socket.on(Socket.EVENT_CONNECTING) {
  logger.debug("------------------------------")
  logger.debug("Connecting to WebSocket server")
  logger.debug("------------------------------")
}

socket.on(Socket.EVENT_CONNECT) {
  logger.debug("------------------------------")
  logger.debug("Connected to WebSocket server!")
  logger.debug("------------------------------")
}

socket.on(Socket.EVENT_CONNECT_TIMEOUT) {
  logger.debug("------------------------------")
  logger.debug("Timeout error when connecting!")
  logger.debug("------------------------------")
}

socket.on(Socket.EVENT_CONNECT_ERROR) {
  logger.debug("------------------------------")
  logger.debug("Error connecting to WebSocket!")
  logger.debug("------------------------------")
}
```

None of those callbacks are called, and nothing is logged, even it logs on the server. Can I somehow enable socket.io logs on the Android devices to perhaps see more information about what's going on?

For reference, this is the server log output when connecting using the v0.6.3 of the client:

```
socket.io:server incoming connection with id XDCLKJfrw0PJdgVxAAAA +29s
socket.io:client connecting to namespace / +1ms
socket.io:namespace adding socket to nsp / +0ms
socket.io:socket socket connected - writing packet +2ms
socket.io:socket joining room XDCLKJfrw0PJdgVxAAAA +0ms
socket.io:client writing packet {"type":0,"nsp":"/"} +0ms
socket.io-parser encoding packet {"type":0,"nsp":"/"} +1ms
socket.io-parser encoded {"type":0,"nsp":"/"} as 0 +0ms
socket.io:socket joined room XDCLKJfrw0PJdgVxAAAA +1ms
```

I feel like this is a bug in the Java client, which is why I'm posting it here. However, I'm stumped that I haven't been able to find anyone else with the same issue. Does anyone here know what's going on? :confused:

**Comment by sindrenm** [comment]
Oh, and I'm using v1.5.0 of the socket.io server!

**Comment by sindrenm** [comment]
And I'm not connecting over SSH, the server URI looks like this: http://<my-ip>:1816. The port is set by us using the following code:

```javascript
import WebSocketServer from 'socket.io';

this._wsServer = new WebSocketServer(1816, {
  pingInterval: 1000,
  pingTimeout: 5000,
});
```

**Comment by varunagarwal315** [comment]
I am using the exact same library version as you and its working fine
for me. All events are being connected, perhaps if you could share
some of the android code

,I could help further.

My gradle looks like this
compile('io.socket:socket.io-client:0.7.0') {
    exclude group: 'org.json', module: 'json'
}

**Comment by sindrenm** [comment]
That's exactly like my Gradle config, @varunagarwal315. Here's my connecting logic as well. Anything else in partucular you're thinking about:

```kotlin
val socket = IO.socket("http://<my-ip>:1816")

socket.connect()
```

**Comment by varunagarwal315** [comment]
connecting code looks fine, why dont you try this to test the connection
further.

Emit some value from android.
*mySocket.emit("test_event" , "ping")*

In your server set up a receiver for that event
*socket.on("test_event", console.log(args[0]) *
and see if the data is being sent. If its successful then modify it to send
a message back to the android client via some other event.

Let me know how this works, code I have written is for nodejs. If you are
using some other server languages, make whatever changes are needed.
At least we will have confirmation whether the connection is setup properly
or not irrespective of whether the default events are not being called.

Thanks,
Varun Agarwal

**Comment by sindrenm** [comment]
But when do I start emitting events from the Android client? I don't know when I'm connected. I suppose I could try and send some events from the game server, at least, see if the Android client receives them.

**Comment by varunagarwal315** [comment]
Yes, do both, on server side

socket.emit("foo", "ping")
socket.on("bar", console.log(args[0])

and android
socket.emit("bar", "pong")
socket.on("foo" , println(args[0[)

The beauty is you can send the event from android whether you are connected
or not. As soon as you get connected the message will be sent as the socket
client waits patiently for a connection to be set via a setTimeout() method.

**Comment by sindrenm** [comment]
It doesn't seem like the Java client deems the connection as open:

```
Connected client!
  socket.io:client writing packet {"type":2,"data":["message","{\"command\":\"TEST PING EVERYTHING\"}"],"nsp":"/"} +2ms
  socket.io-parser encoding packet {"type":2,"data":["message","{\"command\":\"TEST PING EVERYTHING\"}"],"nsp":"/"} +0ms
  socket.io-parser encoded {"type":2,"data":["message","{\"command\":\"TEST PING EVERYTHING\"}"],"nsp":"/"} as 2["message","{\"command\":\"TEST PING EVERYTHING\"}"] +0ms
  socket.io:socket joined room kGgu8ElPlcohNP5YAAAA +0ms
```

This message is never received on the client.

**Comment by varunagarwal315** [comment]
I am not using Kotlin, perhaps there may be an issue in that itself. By the
way, from my understanding, Kotlin is not ready for production. It does not
support a large number of libraries as yet and would advise against it,
unless you are fairly confident in it.

If the server receives a connection then android must have it

**Comment by sindrenm** [comment]
Well, as I mentioned it worked fine with v0.6.3 of the library, so I don't see why this should be any different. I could rewrite this single class to Java, but I'm almost certain it's not going to make a difference. Also, I believe the connection is open, but that the client just doesn't know that it is. Trying to send from the client also doesn't work (i.e. the server doesn't receive it). Do you know if I can enable debug logs from socket.io or something on the client side to perhaps investigate further?

**Comment by varunagarwal315** [comment]
Socket.io has undergone a lot of changes, perhaps it would be best to stick
to v0.6.3 or write a java class for socket as a singleton for v0.7.0. As I
said, Kotlin does have its issues, maybe some changes to the lib is causing
this.

Socket has Manager and Options class but they do not have any feature for
logging. Its more for controlling the timeout, connection attempts, headers
and other stuff.

Sorry, cant help any more.

**Comment by sindrenm** [comment]
Yeah, I'm converting it to a Java class now, though, just to be 100 % certain, but I'm quite sure that's not the issue. I might also just try a new project that doesn't have all of the extra stuff (this project is quite large) and test with both Java and Kotlin there.

Thanks for your help and confirmation that it's working for you. Then there should be something that I can do to fix it. Could you share what version of socket.io your server is running as well, so I'm able to best reproduce your setup? :smile:

**Comment by varunagarwal315** [comment]
Node js : "socket.io": "^1.4.0"
Can share snippets of my code in a couple of hours if you really need it.

What I would suggest is to just clone this
*https://github.com/nkzawa/socket.io-android-chat
<https://github.com/nkzawa/socket.io-android-chat>* , and change the
connecting http address to that of your server. They already got functions
for the onConnection and disconnection events. So you don't even need to
write a separate android code to test it out. This will be the fastest way
IMO to move forward.

**Comment by sindrenm** [comment]
Yeah, thanks! I'm using 1.5.0 on the server, but that should still be compatible, I guess. Again, 1.5.0 works with the 0.6.3 client. :stuck_out_tongue:

**Comment by varunagarwal315** [comment]
Ok clone the repo and test it out, let me know how it goes, if it works,
you got your answer.

**Comment by sindrenm** [comment]
Err, is the socket still cached when calling `IO.socket`, btw? Just realized that my current code relies on that. After I started caching it myself, I began receiving errors as well:

>Error: io.socket.engineio.client.EngineIOException: xhr poll error

I found https://github.com/socketio/socket.io-client-java/issues/250, but in my case it occurs every single time it tries to connect. However, it looks as if it's connected on the server, as mentioned earlier.

**Comment by varunagarwal315** [comment]
Sorry, did not understand your question. Afaik, you have to call
socket.disconnect() manually in android, so the socket instance should
still be running. Sometimes clearing app cache and re-installing the app
can solve things. Hard to say anything if I cannot recreate your error or
see bits of your code to get a better understanding.

**Comment by sindrenm** [comment]
As far as I understand it, calling `IO.socket(serverUri)` should either create a new socket and return that or give back an existing (cached) one if it already exists. My previous code relied on it being cached by the library, but if it wasn't, that would explain why I didn't get any error messages (it would create a new one every time I assigned a callback to it). Now I'm caching it myself and I'm receiving that error message about the XHR polling.

Clearing the app cache now actually resolved the error as well! [Insert every single happy smiley and champagne bottle emoji here.] Thanks! :grinning: 

But it does indeed look like the socket isn't cached anymore for some reason, or that something changed in the way it was caching from 0.6.3 to 0.7.0 and beyond. Oh well, I can easily do the caching on my side, so there's no worry.

Thank you for your quick responses and help, @varunagarwal315! You're the boss. :sunglasses:

**Comment by varunagarwal315** [comment]
No issues at all. I am currently using a Singleton pattern to create one
global socket instance that I can call all throughout the app. I actually
initiate this on a Service.class since I am using socket.io for a chat app.

May I know how are you using your socket instance? and why exactly do you
have a requirement for caching the socket. Also just curious, are you
providing custom socket ids to your users?

When dealing with sockets and even SQLite db, on android, do not use quick
start to install new apk after making minor changes , instead always ensure
the app is off and then install new apk, I learned this the hard way.

**Comment by sindrenm** [comment]
Oh, yeah, I can imagine that would produce a bunch of random hard-to-figure-out errors! Nice to know, I would probably fall into that same trap one time myself. :laughing: 

I'm also using a service, here's the general outline of my code:

```kotlin
class WebSocketService() : Service() {

  private val serviceBinder = ServiceBinder()

  private val socket: Socket
    get() = IO.socket(BuildConfig.GAME_SERVER_URI)

  private lateinit var onOpenedCallback: () -> Unit

  override fun onCreate() {
    socket.on(Socket.EVENT_CONNECT) {
      logger.debug("Connected to WebSocket server!")

      onOpenedCallback()
    }
  }

  override fun onBind(intent: Intent) = serviceBinder

  override fun onUnbind(intent: Intent?): Boolean {
    socket.close()
    return false
  }

  fun connectToServer(onOpenedCallback: () -> Unit) {
    this.onOpenedCallback = onOpenedCallback
    socket.connect()
  }

  fun sendWebSocketPacket(packet: Packet) {
    socket.send(packet.encode())
  }

  inner class ServiceBinder : Binder() {
    val service = this@WebSocketService
  }

}
```

The getter for the socket instance will be called every time I reference the `socket` in the class. If it's not cached by the library, then it'll create a new one and assign the event listeners to that one instead. However, lazy-initializing the object will create it only once (when it's referenced the first time) and then cache that value for later references:

```kotlin
private val socket: Socket by lazy {
  IO.socket(BuildConfig.GAME_SERVER_URI)
}
```

This works like a charm now. :grinning:

**Comment by varunagarwal315** [comment]
My service class looks the same, how i have a separate Singleton class
where I instantiate the socket. Not much of a difference.

**Comment by sindrenm** [comment]
Yeah, but that is basically the difference that matters. You have the singleton class that ensures that only one instance of a socket is in play, I thought that was done by the framework (reading [this blog post][1] and also the [README][2]), but since it doesn't seem to do that anymore (bug?), I ended up putting listeners on random new instances every time I referenced the getter function for the socket, since that method doesn't cache anything. You see what I mean? :smile: 

[1]: http://socket.io/blog/native-socket-io-and-android/#using-socket-in-activity-and-fragment
[2]: https://github.com/socketio/socket.io-client-java#readme
